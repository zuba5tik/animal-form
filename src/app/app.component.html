<body>
  <form [formGroup]="animalForm">
    <mat-card>
      <mat-form-field>
        <input matInput placeholder="Name" type="text" formControlName="name">
      </mat-form-field>
      <p></p>
      <!-- <div [ngSwitch]="animalForm.value.type" > -->
        <!-- <div *ngSwitchCase="'text'"> -->
          <!-- <mat-form-field>
            <input matInput placeholder="Animal Type" type="text" formControlName="type">
          </mat-form-field> -->
        <!-- </div> -->
        <!-- <div *ngSwitchCase="'dropdown'"> -->
            <mat-form-field>
              <mat-select placeholder="Animal Type" formControlName="type">
                  <mat-option *ngFor="let ani of types" [value]="ani.key">
                    {{ani.label}}
                  </mat-option>
              </mat-select>
            </mat-form-field>
          <!-- </div>
      </div> -->
      <div>
        <button mat-raised-button color="accent">Dropdown</button>
      </div>
    </mat-card>
  </form>
  <pre>{{animalForm.value|json}}</pre>
</body>
