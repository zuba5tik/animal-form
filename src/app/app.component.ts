import { Component, OnInit } from '@angular/core';
import { FormGroup, FormControl, FormBuilder } from '@angular/forms';

// export interface FormField {
//   label: string;
//   type: string;
// }
// export interface InputType {
//   type: string;
// }
// export const inputTypes: InputType[] = [
//   {type: 'text'}, {type: 'select'}
// ];
// export function getFormField(formField: FormField = null): FormGroup {
//   return new FormGroup({
//     label: new FormControl(formField ? formField.label : 'Label'),
//     type: new FormControl(formField ? formField.type : inputTypes[0].type),
//   });
// }

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {
  title = 'animal-app';
  animalForm: FormGroup;
  animalType = new FormGroup({
    typeLabel: new FormControl(),
    typeKey: new FormControl(),
  });
  constructor(private fb: FormBuilder) {
    this.animalForm = this.fb.group({
      animalName: [],
      animalType: [],
      // animalType: this.animalType = this.fb.group({
      //   typeLabel: [],
      //   typeKey: [],
      // })
    });
  }

  inputDropDown() {
    // this.animalType.addControl('typeKey', new FormControl());
    this.animalForm.reset();
    this.animalForm.removeControl('animalType');
    this.animalForm.registerControl('animalType', new FormGroup({typeLabel: new FormControl(), typeKey: new FormControl()}));
    // this.animalType.addControl('typeLabel', new FormControl());
    // this.animalType.addControl('typeKey', new FormControl());
    this.inputType = 'select';
  }

  inputText() {
    this.animalForm.reset();
    this.animalForm.removeControl('animalType');
    this.animalForm.addControl('animalType', new FormControl());
    this.inputType = 'text';
  }

  // getField() {
  //   this.types = [
  //     {
  //       'key': this.animalType.typeKey,
  //       'label': this.animalType.typeLabel,
  //     },
  //     ...this.types,
  //   ];
  // }

  // tslint:disable-next-line:member-ordering
  inputType = 'text';

  // tslint:disable-next-line:member-ordering
  types = [
    {'key': 'dog-0', 'label': 'Dog'},
    {'key': 'cat-1', 'label': 'Cat'}
  ];

  // GetControl(form: FormGroup, field: string) {
  //   // tslint:disable-next-line:whitespace
  //   const fieldType = document.querySelector('input[name="'+field+'"]');
  //   if (fieldType.getAttribute('type') === 'text') {
  //     console.log('nice');
  //   }
  //   return form.get(field);
  // }

  // fieldType = [
  //   {type: 'text', placeholder: 'Animal Type text'},
  //   {type: 'select', placeholder: 'Animal Type select'}
  // ];
}
