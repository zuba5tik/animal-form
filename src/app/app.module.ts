import { BrowserModule } from '@angular/platform-browser';
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
import { HttpClientModule } from '@angular/common/http';
import {MatCheckboxModule,
  MatButtonModule,
  MatInputModule,
  MatAutocompleteModule,
  MatDatepickerModule,
  MatFormFieldModule,
  MatRadioModule,
  MatSelectModule,
  MatSliderModule,
  MatSlideToggleModule,
  MatMenuModule,
  MatSidenavModule,
  MatToolbarModule,
  MatListModule,
  MatGridListModule,
  MatCardModule,
  MatStepperModule,
  MatTabsModule,
  MatExpansionModule,
  MatButtonToggleModule,
  MatChipsModule,
  MatIconModule,
  MatProgressSpinnerModule,
  MatProgressBarModule,
  MatDialogModule,
  MatTooltipModule,
  MatSnackBarModule,
  MatTableModule,
  MatSortModule,
MatPaginatorModule} from '@angular/material';

import { AppComponent } from './app.component';
import { AppTestComponent } from './app-test/app-test.component';
import { LoginFormComponent } from './login-form/login-form.component';

@NgModule({
  declarations: [
    AppComponent,
    AppTestComponent,
    LoginFormComponent
  ],
  imports: [
    BrowserModule,
    HttpClientModule,
    ReactiveFormsModule,
    BrowserAnimationsModule,
    MatCheckboxModule,
    MatCheckboxModule,
    MatButtonModule,
    MatInputModule,
    MatAutocompleteModule,
    MatDatepickerModule,
    MatFormFieldModule,
    MatRadioModule,
    MatSelectModule,
    MatSliderModule,
    MatSlideToggleModule,
    MatMenuModule,
    MatSidenavModule,
    MatToolbarModule,
    MatListModule,
    MatGridListModule,
    MatCardModule,
    MatStepperModule,
    MatTabsModule,
    MatExpansionModule,
    MatButtonToggleModule,
    MatChipsModule,
    MatIconModule,
    MatProgressSpinnerModule,
    MatProgressBarModule,
    MatDialogModule,
    MatTooltipModule,
    MatSnackBarModule,
    MatTableModule,
    MatSortModule,
    MatPaginatorModule
  ],
  providers: [],
  bootstrap: [AppComponent],
  schemas: [ CUSTOM_ELEMENTS_SCHEMA ],
})
export class AppModule { }
