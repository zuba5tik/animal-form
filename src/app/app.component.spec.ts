import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { AppComponent } from './app.component';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';
import { text } from '@angular/core/src/render3/instructions';

describe('Animal form', () => {
  let fixture: ComponentFixture<AppComponent>;
  let component: AppComponent;

  beforeEach(async() => {
    TestBed.configureTestingModule({
      imports: [ FormsModule, ReactiveFormsModule],
      declarations: [AppComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
  });

  it('is component defined', () => {
    expect(component).toBeDefined();
  });

  it('is form invalid when empty', () => {
    let petName = component.animalForm.controls['animalName'];
    petName.setValue(null);
    // let petLabel = component.animalForm.controls['typeLabel'];
    // petLabel.setValue(null);
    expect(component.animalForm.controls['animalName']).toMatchSnapshot();
    // expect(component.animalForm.controls['typeLabel'].valid).toBeFalsy();
  });

  it('should click', async(() => {
    spyOn(component, 'inputDropDown');

    let button = fixture.debugElement.nativeElement.querySelector('button');
    button.click();

    fixture.whenStable().then(() => {
      expect(component.inputDropDown).toHaveBeenCalled();
    });
  }));

  it('inputDropDown should work', () => {
    component.inputType = 'text';
    component.inputDropDown();

    expect(component.inputType).toMatchSnapshot();
    expect(component.animalForm).toMatchSnapshot();
  });

  it('shoud spy', () => {
    let petName = component.animalForm.controls['animalName'];
    petName.setValue('123');
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
