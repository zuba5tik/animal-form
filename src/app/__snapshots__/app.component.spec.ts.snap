// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animal form inputDropDown should work 1`] = `"select"`;

exports[`Animal form inputDropDown should work 2`] = `
FormGroup {
  "_onCollectionChange": [Function],
  "_onDisabledChange": Array [],
  "asyncValidator": null,
  "controls": Object {
    "animalName": FormControl {
      "_onChange": Array [],
      "_onCollectionChange": [Function],
      "_onDisabledChange": Array [],
      "_parent": [Circular],
      "_pendingChange": false,
      "_pendingDirty": false,
      "_pendingTouched": false,
      "_pendingValue": null,
      "asyncValidator": null,
      "errors": Object {
        "required": true,
      },
      "pristine": true,
      "status": "INVALID",
      "statusChanges": EventEmitter {
        "__isAsync": false,
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
      "touched": false,
      "validator": [Function],
      "value": null,
      "valueChanges": EventEmitter {
        "__isAsync": false,
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
    },
    "animalType": FormGroup {
      "_onCollectionChange": [Function],
      "_onDisabledChange": Array [],
      "_parent": [Circular],
      "asyncValidator": null,
      "controls": Object {
        "typeKey": FormControl {
          "_onChange": Array [],
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingValue": null,
          "asyncValidator": null,
          "errors": null,
          "pristine": true,
          "status": "VALID",
          "statusChanges": EventEmitter {
            "__isAsync": false,
            "_isScalar": false,
            "closed": false,
            "hasError": false,
            "isStopped": false,
            "observers": Array [],
            "thrownError": null,
          },
          "touched": false,
          "validator": null,
          "value": null,
          "valueChanges": EventEmitter {
            "__isAsync": false,
            "_isScalar": false,
            "closed": false,
            "hasError": false,
            "isStopped": false,
            "observers": Array [],
            "thrownError": null,
          },
        },
        "typeLabel": FormControl {
          "_onChange": Array [],
          "_onCollectionChange": [Function],
          "_onDisabledChange": Array [],
          "_parent": [Circular],
          "_pendingValue": null,
          "asyncValidator": null,
          "errors": null,
          "pristine": true,
          "status": "VALID",
          "statusChanges": EventEmitter {
            "__isAsync": false,
            "_isScalar": false,
            "closed": false,
            "hasError": false,
            "isStopped": false,
            "observers": Array [],
            "thrownError": null,
          },
          "touched": false,
          "validator": null,
          "value": null,
          "valueChanges": EventEmitter {
            "__isAsync": false,
            "_isScalar": false,
            "closed": false,
            "hasError": false,
            "isStopped": false,
            "observers": Array [],
            "thrownError": null,
          },
        },
      },
      "errors": null,
      "pristine": true,
      "status": "VALID",
      "statusChanges": EventEmitter {
        "__isAsync": false,
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
      "touched": false,
      "validator": null,
      "value": Object {
        "typeKey": null,
        "typeLabel": null,
      },
      "valueChanges": EventEmitter {
        "__isAsync": false,
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
    },
  },
  "errors": null,
  "pristine": true,
  "status": "INVALID",
  "statusChanges": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
  "touched": false,
  "validator": null,
  "value": Object {
    "animalName": null,
  },
  "valueChanges": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
}
`;

exports[`Animal form is form invalid when empty 1`] = `
FormControl {
  "_onChange": Array [],
  "_onCollectionChange": [Function],
  "_onDisabledChange": Array [],
  "_parent": FormGroup {
    "_onCollectionChange": [Function],
    "_onDisabledChange": Array [],
    "asyncValidator": null,
    "controls": Object {
      "animalName": [Circular],
      "animalType": FormGroup {
        "_onCollectionChange": [Function],
        "_onDisabledChange": Array [],
        "_parent": [Circular],
        "asyncValidator": null,
        "controls": Object {
          "typeKey": FormControl {
            "_onChange": Array [],
            "_onCollectionChange": [Function],
            "_onDisabledChange": Array [],
            "_parent": [Circular],
            "_pendingValue": null,
            "asyncValidator": null,
            "errors": null,
            "pristine": true,
            "status": "VALID",
            "statusChanges": EventEmitter {
              "__isAsync": false,
              "_isScalar": false,
              "closed": false,
              "hasError": false,
              "isStopped": false,
              "observers": Array [],
              "thrownError": null,
            },
            "touched": false,
            "validator": null,
            "value": null,
            "valueChanges": EventEmitter {
              "__isAsync": false,
              "_isScalar": false,
              "closed": false,
              "hasError": false,
              "isStopped": false,
              "observers": Array [],
              "thrownError": null,
            },
          },
          "typeLabel": FormControl {
            "_onChange": Array [],
            "_onCollectionChange": [Function],
            "_onDisabledChange": Array [],
            "_parent": [Circular],
            "_pendingValue": "",
            "asyncValidator": null,
            "errors": Object {
              "required": true,
            },
            "pristine": true,
            "status": "INVALID",
            "statusChanges": EventEmitter {
              "__isAsync": false,
              "_isScalar": false,
              "closed": false,
              "hasError": false,
              "isStopped": false,
              "observers": Array [],
              "thrownError": null,
            },
            "touched": false,
            "validator": [Function],
            "value": "",
            "valueChanges": EventEmitter {
              "__isAsync": false,
              "_isScalar": false,
              "closed": false,
              "hasError": false,
              "isStopped": false,
              "observers": Array [],
              "thrownError": null,
            },
          },
        },
        "errors": null,
        "pristine": true,
        "status": "INVALID",
        "statusChanges": EventEmitter {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "touched": false,
        "validator": null,
        "value": Object {
          "typeKey": null,
          "typeLabel": "",
        },
        "valueChanges": EventEmitter {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
    },
    "errors": null,
    "pristine": true,
    "status": "INVALID",
    "statusChanges": EventEmitter {
      "__isAsync": false,
      "_isScalar": false,
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "touched": false,
    "validator": null,
    "value": Object {
      "animalName": null,
      "animalType": Object {
        "typeKey": null,
        "typeLabel": "",
      },
    },
    "valueChanges": EventEmitter {
      "__isAsync": false,
      "_isScalar": false,
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
  },
  "_pendingValue": null,
  "asyncValidator": null,
  "errors": Object {
    "required": true,
  },
  "pristine": true,
  "status": "INVALID",
  "statusChanges": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
  "touched": false,
  "validator": [Function],
  "value": null,
  "valueChanges": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
}
`;

exports[`Animal form shoud spy 1`] = `
<app-root
  animalForm={[Function FormGroup]}
  animalType={[Function FormGroup]}
  fb={[Function FormBuilder]}
  inputType={[Function String]}
  title={[Function String]}
  types={[Function Array]}
>
  <body>
    <form
      class="ng-untouched ng-pristine ng-invalid"
      ng-reflect-form="[object Object]"
      novalidate=""
    >
      <input
        class="a ng-untouched ng-pristine ng-invalid"
        formcontrolname="animalName"
        matinput=""
        ng-reflect-name="animalName"
        placeholder="Name"
        type="text"
      />
      <p />
      
      
      <div
        class="ng-untouched ng-pristine ng-invalid"
        formgroupname="animalType"
        ng-reflect-name="animalType"
      >
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="typeKey"
          matinput=""
          ng-reflect-name="typeKey"
          placeholder="Key"
          type="text"
        />
        <input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="typeLabel"
          matinput=""
          ng-reflect-name="typeLabel"
          placeholder="Label"
          type="text"
        />
      </div>
      <div
        class="a"
      >
        <button
          color="accent"
          mat-raised-button=""
        >
          Change field type
        </button>
      </div>
    </form>
    <pre>
      {
    "animalName": "123",
    "animalType": {
      "typeLabel": "",
      "typeKey": null
    }
  }
    </pre>
  </body>
</app-root>
`;
